trie/database_wrap.go

```diff
@@ -0,0 +1,267 @@
+	if db.preimages != nil {
+		db.preimages.commit(true)
+	}
+	return db.backend.Commit(root, report)
+}
+
+// Size returns the storage size of dirty trie nodes in front of the persistent
+// database and the size of cached preimages.
+func (db *Database) Size() (common.StorageSize, common.StorageSize) {
+	var (
+		storages  common.StorageSize
+		preimages common.StorageSize
+	)
+	storages = db.backend.Size()
+	if db.preimages != nil {
+		preimages = db.preimages.size()
+	}
+	return storages, preimages
+}
+
+// Initialized returns an indicator if the state data is already initialized
+// according to the state scheme.
+func (db *Database) Initialized(genesisRoot common.Hash) bool {
+	return db.backend.Initialized(genesisRoot)
+}
+
+// Scheme returns the node scheme used in the database.
+func (db *Database) Scheme() string {
+	return db.backend.Scheme()
+}
+
+// Close flushes the dangling preimages to disk and closes the trie database.
+// It is meant to be called when closing the blockchain object, so that all
+// resources held can be released correctly.
+func (db *Database) Close() error {
+	if db.preimages != nil {
+		db.preimages.commit(true)
+	}
+	return db.backend.Close()
+}
+
+// saveCache saves clean state cache to given directory path
+// using specified CPU cores.
+func (db *Database) saveCache(dir string, threads int) error {
+	if db.cleans == nil {
+		return nil
+	}
+	log.Info("Writing clean trie cache to disk", "path", dir, "threads", threads)
+
+	start := time.Now()
+	err := db.cleans.SaveToFileConcurrent(dir, threads)
+	if err != nil {
+		log.Error("Failed to persist clean trie cache", "error", err)
+		return err
+	}
+	log.Info("Persisted the clean trie cache", "path", dir, "elapsed", common.PrettyDuration(time.Since(start)))
+	return nil
+}
+
+// SaveCache atomically saves fast cache data to the given dir using all
+// available CPU cores.
+func (db *Database) SaveCache(dir string) error {
+	return db.saveCache(dir, runtime.GOMAXPROCS(0))
+}
+
+// SaveCachePeriodically atomically saves fast cache data to the given dir with
+// the specified interval. All dump operation will only use a single CPU core.
+func (db *Database) SaveCachePeriodically(dir string, interval time.Duration, stopCh <-chan struct{}) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			db.saveCache(dir, 1)
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+// Cap iteratively flushes old but still referenced trie nodes until the total
+// memory usage goes below the given threshold. The held pre-images accumulated
+// up to this point will be flushed in case the size exceeds the threshold.
+//
+// It's only supported by hash-based database and will return an error for others.
+func (db *Database) Cap(limit common.StorageSize) error {
+	hdb, ok := db.backend.(*hashdb.Database)
+	if !ok {
+		return errors.New("not supported")
+	}
+	if db.preimages != nil {
+		db.preimages.commit(false)
+	}
```

Comment:
For non-hash based backends the preimages will not be commited. Is this by design, or should this preimage commit be moved before the check? Feels that if the preimage commit is not needed for path based scheme, it should not be needed for hash based either. But if it's needed for hash, I'd guess it's needed for path too?

Reply:
> For non-hash based backends the preimages will not be commited. Is this by design, or should this preimage commit be moved before the check?

It's by design. Because `Cap` is not used in path-based at all and only called in hash-based. I deliberately add the check at the beginning.

Reply:
I agree it's very hacky since hash-based and path-based have some semantics differences.

`Reference`, `Dereference` are used to link the account trie and storage trie, it's not required in path-based.

`Cap` is used to control the memory consumption by in-memory nodes and evict out some oldest nodes if necessary. In PBSS we can implement the similar things by evicting the oldest layers. Whereas PBSS already limits the layers by number(128), I am not sure we also need the `Cap`.

Besides, pbss also has some it's own notions, like `Revert`, `Revertable`, etc, which are not existent in hash-based for sure. These specific functions are meant to be called in the corresponding scheme only.

Reply:
Fair enough